packages/local/XHttp/src/XHttp.cpp

```diff
@@ -0,0 +1,126 @@
+#include <services/local/XHttp.hpp>
+
+#include <psibase/dispatch.hpp>
+#include <psibase/webServices.hpp>
+
+using namespace psibase;
+using namespace LocalService;
+using namespace SystemService;
+
+using Temporary = TemporaryTables<PendingRequestTable>;
+
+extern "C" [[clang::export_name("recv")]] void recv()
+{
+   auto act       = getCurrentActionView();
+   act->sender()  = XHttp::service;
+   act->service() = HttpServer::service;
+   act->method()  = MethodNumber{"recv"};
+   psibase::call(act.data(), act.size());
+}
+
+void XHttp::send(std::int32_t socket, psio::view<const std::vector<char>> data)
+{
+   check(getSender() == HttpServer::service, "Wrong sender");
+   check(socket == producer_multicast, "Cannot call send on this socket");
+   psibase::socketSend(socket, data);
+}
+
+void XHttp::autoClose(std::int32_t socket, bool value)
+{
+   auto sender = getSender();
+   PSIBASE_SUBJECTIVE_TX
+   {
+      auto requests = Subjective{}.open<PendingRequestTable>();
+      auto owned    = Temporary{}.open<PendingRequestTable>();
+
+      auto from = value ? &requests : &owned;
+      auto to   = value ? &owned : &requests;
+      auto row  = from->get(socket);
```

Comment:
I would probably expect this function to be idempotent

Reply:
In practice it doesn't matter much, because the caller has to keep autoClose in sync with its own tables.